src/bin/pg_dump/dump_babel_utils.c

```diff
@@ -46,3 +68,109 @@ bbf_selectDumpableCast(CastInfo *cast)
 			 strcmp(tTypeInfo->dobj.name, "varchar") == 0))
 		cast->dobj.dump = DUMP_COMPONENT_NONE;
 }
+
+/*
+ * bbf_fixTableTypeDependency:
+ * By default function gets dumped before the template table of T-SQL
+ * table type(one of the datatype of function's arguments) which is
+ * because there is no dependency between function and underlying
+ * template table. Which is fine in normal case but becomes problematic
+ * during restore. Fix this by adding function's dependency on 
+ * template table. 
+ */
+void
+bbf_fixTableTypeDependency(Archive *fout, DumpableObject *func, DumpableObject *tabletype)
+{
+	FuncInfo  *funcInfo = (FuncInfo *) func;
+	TypeInfo  *typeInfo = (TypeInfo *) tabletype;
+	TableInfo *tytable;
+
+	/* skip auto-generated array types and non-pltsql functions */
+	if (typeInfo->isArray ||
+		!OidIsValid(typeInfo->typrelid) ||
+		strcmp(get_language_name(fout, funcInfo->lang), "pltsql") != 0)
+		return;
+
+	tytable = findTableByOid(typeInfo->typrelid);
+
+	if (tytable == NULL)
+		return;
+
+	/* Add function's dependency on template table */
+	addObjectDependency(func, tytable->dobj.dumpId);
+}
+
+/*
+ * bbf_is_tsqltabletype:
+ * Returns true if given table is a template table for
+ * underlying T-SQL table-type, false otherwise.
+ */
+bool
+bbf_is_tsqltabletype(Archive *fout, const TableInfo *tbinfo)
```

Comment:
Is there any possibility it returns true for non-bbf objects?

Reply:
This function will return `TRUE` only for T-SQL table-type as a table can not have a dependency on it's row type (dummy type) in PG, unless it's a standalone composite type (which PG will dump as a type and not a table). So we are good here.